---
title: "popGenome_analysis"
author: "Christian Parobek"
date: "01/20/2015"
output: html_document
---


Might need to split up my genome into chromosomes:
```{r, engine='bash', cache=TRUE}
# For Pv
for i in 01 02 03 04 05 06 07 08 09 10 11 12 13 14
do

grep "^#\|Pv_Sal1_chr$i" /proj/julianog/users/ChristianP/cambodiaWGS/variants/good69.pass.vcf | grep -v "AAKM" > test/vcf/chr$i.vcf
  
grep "^#\|Pv_Sal1_chr$i" /proj/julianog/refs/PvSAL1_v13.0/PlasmoDB-13.0_PvivaxSal1.gff | grep -v "##FASTA\|>Pv_Sal1_chr$i\|AAKM"> test/gff/chr$i.gff

done

# For Pf
for i in 01 02 03 04 05 06 07 08 09 10 11 12 13 14
do

grep "^#\|^Pf3D7_$i\_v3" /proj/julianog/users/ChristianP/cambodiaWGS/PopGenome/pf_whole_genome/our_goods_UG.pass.vcf > /proj/julianog/users/ChristianP/cambodiaWGS/PopGenome/pf_whole_genome/vcf/chr$i.vcf

grep "^#\|^Pf3D7_$i\_v3" /proj/julianog/refs/Pf3D7_v13.0/PlasmoDB-13.0_Pfalciparum3D7.gff > /proj/julianog/users/ChristianP/cambodiaWGS/PopGenome/pf_whole_genome/gff/chr$i.gff 

done
```

As it turns out, PopGenome infers the chromosome length from the position of the final VCF entry. So, for example, on Chromosome 1, the final entry for Pv_Sal1_chr01 is at 800738, so that's how long the `@n.sites` command says the chromosome is. That might be OK.

Also, when I import a multiVCF file using readData(), any ALT alleles that are fixed in the population contained within the multiVCF are dropped from the `GENOME.class` object that PopGenome produces, so the number of variant entries might be a little lower in the GENOME.class object than in the original VCF file.

When importing the GFF, watch out for any spurious lines after the GFF data section. For example, I may have to delete lines that are comprised of "##FASTA"" or ">Pv_Sal1_chr01""

Next, read in the data:
```{r}
## Load the PopGenome library
library(PopGenome)
library(stringr)

## Read in the VCFs and GFFs
pv <- readData("pv_whole_genome/vcf/", format="VCF", gffpath = "pv_whole_genome/gff/")

## Add REF info for Syn/NonSyn Sites
pv <- set.synnonsyn(pv, ref.chr=c("pv_whole_genome/ref/chr01.fasta", "pv_whole_genome/ref/chr02.fasta","pv_whole_genome/ref/chr03.fasta","pv_whole_genome/ref/chr04.fasta","pv_whole_genome/ref/chr05.fasta","pv_whole_genome/ref/chr06.fasta","pv_whole_genome/ref/chr07.fasta","pv_whole_genome/ref/chr08.fasta","pv_whole_genome/ref/chr09.fasta","pv_whole_genome/ref/chr10.fasta","pv_whole_genome/ref/chr11.fasta","pv_whole_genome/ref/chr12.fasta","pv_whole_genome/ref/chr13.fasta","pv_whole_genome/ref/chr14.fasta"))
  # 1 is syn
  # 0 is nonsyn

## Define populations: BB, KP, and OM
indivs <- get.individuals(pv)
pvPops <- set.populations(pv, list(
  indivs[[1]][1:69][grep("BB", indivs[[1]][1:length(indivs[[1]])])], 
  indivs[[1]][1:69][grep("KP", indivs[[1]][1:length(indivs[[1]])])], 
  indivs[[1]][1:69][grep("OM", indivs[[1]][1:length(indivs[[1]])])]))

## Split data into genes
onePop_genes <- splitting.data(pv, subsites="gene", whole.data = FALSE)
  # Specify whole.data = FALSE because I don't want to concatenate regions

## Extracting gene IDs from the gff INFO field
## Need to use stringr's str_extract function to actually get the gene ID out of the big long string of INFO it returns
chr01.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr01.gff", chr="Pv_Sal1_chr01", extract.gene.names=TRUE), "PVX_[0-9]+")
chr02.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr02.gff", chr="Pv_Sal1_chr02", extract.gene.names=TRUE), "PVX_[0-9]+")
chr03.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr03.gff", chr="Pv_Sal1_chr03", extract.gene.names=TRUE), "PVX_[0-9]+")
chr04.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr04.gff", chr="Pv_Sal1_chr04", extract.gene.names=TRUE), "PVX_[0-9]+")
chr05.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr05.gff", chr="Pv_Sal1_chr05", extract.gene.names=TRUE), "PVX_[0-9]+")
chr06.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr06.gff", chr="Pv_Sal1_chr06", extract.gene.names=TRUE), "PVX_[0-9]+")
chr07.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr07.gff", chr="Pv_Sal1_chr07", extract.gene.names=TRUE), "PVX_[0-9]+")
chr08.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr08.gff", chr="Pv_Sal1_chr08", extract.gene.names=TRUE), "PVX_[0-9]+")
chr09.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr09.gff", chr="Pv_Sal1_chr09", extract.gene.names=TRUE), "PVX_[0-9]+")
chr10.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr10.gff", chr="Pv_Sal1_chr10", extract.gene.names=TRUE), "PVX_[0-9]+")
chr11.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr11.gff", chr="Pv_Sal1_chr11", extract.gene.names=TRUE), "PVX_[0-9]+")
chr12.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr12.gff", chr="Pv_Sal1_chr12", extract.gene.names=TRUE), "PVX_[0-9]+")
chr13.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr13.gff", chr="Pv_Sal1_chr13", extract.gene.names=TRUE), "PVX_[0-9]+")
chr14.ids <- str_extract(get_gff_info(gff.file="pv_whole_genome/gff/chr14.gff", chr="Pv_Sal1_chr14", extract.gene.names=TRUE), "PVX_[0-9]+")

## Make one big list out of all the little lists of gene IDs
## Is there a more elegant way to do all this?
chr.all.ids <- unlist(list(chr01.ids, chr02.ids, chr03.ids, chr04.ids, chr05.ids, chr06.ids, chr07.ids, chr08.ids, chr09.ids, chr10.ids, chr11.ids, chr12.ids, chr13.ids, chr14.ids))

## Assign the @region.names slot with the new names - YEAH!!
onePop_genes@region.names <- chr.all.ids
```

Calculate Tajima's D for all three Pv populations: OM, BB, and KP. Note that this might not be a fair calculation at all, since we probably shouldn't be doing Tajima's D on small populations like BB and KP.
```{r}
## Calculate Tajima's D by gene for BB+OM+KP as one population
onePop_genes <- neutrality.stats(onePop_genes)

## What if we want to look at Tajima'D by position?
new <- cbind(onePop_genes@Tajima.D, 1:length(onePop_genes@Tajima.D))
plot(new[,1] ~ new[,2])

## Calculate Tajima's D by gene for BB+OM+KP as three separate populations
threePops_genes <- splitting.data(threePops, subsites="gene")
threePops_genes <- neutrality.stats(threePops_genes)
    # Genes for which it gives a NA don't have any variants
    # Could be because they are truly invariant or because they were masked genes
    # Need to confirm that the Tajima's D values I get here are same as calculated in DnaSP!!!!!> 

plot(density(threePops_genes@Tajima.D[,1], na.rm=TRUE), xlim=c(-3,3), lty=2, lwd=2, main="Whole-Genome Genewise Tajima's D: BB, KP, and OM", xlab="Tajima's D", ylab="Kernel Density") # Plot BB
lines(density(threePops_genes@Tajima.D[,2], na.rm=TRUE), lty=3, lwd=2) # Plot KP (??)
lines(density(threePops_genes@Tajima.D[,3], na.rm=TRUE), lty=4, lwd=2) # Plot OM (??)
lines(density(onePop_genes@Tajima.D, na.rm=TRUE), lwd=2) # Plot all three pops together
legend("topright", c("Combined", "BB", "KP", "OM"), lty=c(1,2,3,4), lwd=2)
```

Calculate Tajima's D by gene vs Tajima's D by exon for the OM population:
```{r}
## Calculate Tajima's D by gene for BB+OM+KP as three separate populations
threePops_exons <- splitting.data(threePops, subsites="exon")
threePops_exons <- neutrality.stats(threePops_exons)

plot(density(threePops_exons@Tajima.D[,3], na.rm=TRUE), xlim=c(-3,3), lty=5, lwd=2, main="Whole-Genome Tajima's D: Genes vs. Exons", xlab="Tajima's D", ylab="Kernel Density") # Plot OM exons
lines(density(threePops_genes@Tajima.D[,3], na.rm=TRUE), lty=1, lwd=2) # Plot OM genes

legend("topright", c("Genes", "Exons"), lty=c(1,5), lwd=2)

## Could also look at distribution of TajD in merozoite vs non-merozoite exons
```

Looking for selective sweeps:
```{r}
## Looking for selective sweeps??
## CLR might be more robust to demography, etc (Neilsen 2005)
## Use an FDR to determine which of these are significant

## CLR
## create global set
onePop <- detail.stats(onePop)
freq <- onePop@region.stats@minor.allele.freqs[[1]]
freq.table <- list()
freq.table[[1]] <- table(freq)
## define the region of interest
onePop_genes <- splitting.data(data, subsites="gene")
## calculate CLR
onePop_genes <- sweeps.stats(onePop_genes, freq.table=freq.table)

## Plot CLR Histogram
hist(onePop_genes@CLR)
## Plot CLR by Position
old <- cbind(onePop_genes@CLR, 1:length(onePop_genes@CLR))
plot(old[,1] ~ old[,2])

## Plot CL Histogram
hist(onePop_genes@CL)
## Plot CL by Position
old <- cbind(onePop_genes@CL, 1:length(onePop_genes@CL))
plot(old[,1] ~ old[,2])

```

And doing Fst by Gene:
```{r}
threePops_genes <- F_ST.stats(threePops_genes)
threePops_genes <- detail.stats(threePops_genes, site.FST = TRUE) 
  ## This actually calculates the Fst values that go in the @region.stats@site.FST slot
  ## Do I need to specify populations?

## Plot Fst for a single gene
plot(threePops_genes@region.stats@site.FST[[5219]])

## Histogram of Fsts for all genes
fsts <- lapply(threePops_genes@region.stats@site.FST, FUN = mean)
hist(as.numeric(fsts))

## Plot Fst by Position
plot(as.numeric(fsts))

```

And doing Fst by Exon:
```{r}

threePops_exons <- F_ST.stats(threePops_exons)
threePops_exons <- detail.stats(threePops_exons, site.FST = TRUE) 
  ## This actually calculates the Fst values that go in the @region.stats@site.FST slot
  ## Do I need to specify populations?

## Plot Fst for a single gene
plot(threePops_exons@region.stats@site.FST[[9999]])

## Histogram of Fsts for all exons
fsts <- lapply(threePops_exons@region.stats@site.FST, FUN = mean)
hist(as.numeric(fsts))

## Plot Fst by Position
plot(as.numeric(fsts))
```



Coalescent simulations using `ms` (Hudson 2002).
Set params for coal sims:
http://rpackages.ianhowson.com/cran/PopGenome/man/test.params-class.html
```{r}
## OK maybe want to do coalescent simulations for every gene... locus-by-locus
## Then get 5000 some p-values
## Then do an FDR to find the significant ones

## Seems like I need to read in the data by chromosome
## Wish there were a better way to do this
chr01 <- readData("chr01/vcf/", format="VCF", gffpath = "chr01/gff/", ) 
chr02 <- readData("chr02/vcf/", format="VCF", gffpath = "chr02/gff/", ) 


chr_name <- readData("chr01/vcf/", format="VCF", gffpath = "chr01/gff/", )


## Define function to get coalescent simulations of each gene
## Will need to update/rerun this once I'm using msms, accounting for selection, etc
### SOOOOOOOOO FRUSTRATING!!!
coal.sims <- function(chr_name, index){

  chr_genes <- splitting.data(chr_name, subsites="gene")
  start <- as.numeric(str_extract(chr_genes@region.names[index], "[0-9]+")) ## Get start position
  end <- as.numeric(str_extract(chr_genes@region.names[index], " [0-9]+")) ## Get end position
  gene <- splitting.data(chr_name, positions=list(start:end), type=2) # type=2 indicates it's chr position
  gene <- neutrality.stats(gene)
  gene_coal <- MS(gene, thetaID="Tajima", neutrality=TRUE, niter=500)
   
  ## Get the Tajima's D simulations
  sims <- MS_getStats(gene_coal)[,1]
  obs <- gene_coal@obs.val[[1]][,1] # Get the observed Tajima's D
  ## p-value for balancing selection (pos Taj D)
  bsp <- length(sims[sims > obs])/length(sims)
  
return(bsp)
}






chr01 <- readData("chr01/vcf/", format="VCF", gffpath = "chr01/gff/", ) 
gene <- splitting.data(chr01, positions=list(2350:4494), type=2)
gene <- neutrality.stats(gene)
gene_coales <- MS(gene, thetaID="Tajima", neutrality=TRUE, niter=20)



  chr_genes <- splitting.data(chr01, subsites="gene")
  start <- as.numeric(str_extract(chr_genes@region.names[1], "[0-9]+")) ## Get start position
  end <- as.numeric(str_extract(chr_genes@region.names[1], " [0-9]+")) ## Get end position
  

  gene <- splitting.data(chr01, positions=list(6000:7000), type=2) # type=2 indicates it's chr position
  gene <- neutrality.stats(gene)
  gene_coal <- MS(gene, thetaID="Tajima", neutrality=TRUE, niter=20)
   
  ## Get the Tajima's D simulations
  sims <- MS_getStats(gene_coales)[,1]
  obs <- gene_coales@obs.val[[1]][,1] # Get the observed Tajima's D
  ## p-value for balancing selection (pos Taj D)
  bsp <- length(sims[sims > obs])/length(sims)









## Get a single gene
start <- as.numeric(str_extract(chr01_genes@region.names[1], "[0-9]+")) ## Get start position
end <- as.numeric(str_extract(chr01_genes@region.names[1], " [0-9]+")) ## Get end position
gene1 <- splitting.data(data, positions=list(start:end), type=2) # type=2 indicates it's chr position
gene1 <- neutrality.stats(gene1)
gene1_coal <- MS(gene1, thetaID="Tajima", neutrality=TRUE, niter=500)

## Get the Tajima's D simulations and plot
sims <- MS_getStats(gene1_coal)[,1]
plot(density(sims)) # plot the sims
obs <- gene_coales@obs.val[[1]][,1] # Get the observed Tajima's D
abline(v = obs)

## p-value for balancing selection (pos Taj D)
length(sims[sims > obs])/length(sims)
## p-value for directional selection (neg Taj D)
length(sims[sims < obs])/length(sims)





## How to view the results. What's the diff between the following?
MS_getStats(gene1_coal)[,1]
coal@average[[1]]


## Can go real deep in these datastructure
coal@locus[[14]]@quantiles[[1]]



```


## Code for circos figure
```{r}
chr01 <- readData("chr01/vcf/", format="VCF", gffpath = "chr01/gff/", ) 
chr01_genes <- splitting.data(chr01, subsites="gene")

chr01_genes <- neutrality.stats(chr01_genes) # Calculate Tajima's D
d <- cbind(chr01_genes@region.names, chr01_genes@Tajima.D) # Make TajD Table
write.table(d, file="test_taj.txt") # Output TajD Table

chr01_genes <- diversity.stats(chr01_genes, pi=TRUE) # Calculate Pi
d <- cbind(chr01_genes@region.names, chr01_genes@Pi) # Make Pi Table
write.table(d, file="test_pi.txt") # Output TajD Table

chr01 <- detail.stats(chr01) # Doing CLR... but maybe should figure out how to do it by sliding window rather than by gene
freq <- chr01@region.stats@minor.allele.freqs[[1]]
freq.table <- list()
freq.table[[1]] <- table(freq)
## define the region of interest
chr01_genes <- splitting.data(chr01, subsites="gene")
## calculate CLR
chr01_genes <- sweeps.stats(chr01_genes, freq.table=freq.table)
chr01_genes@CLR
d <- cbind(chr01_genes@region.names, chr01_genes@CLR) # Make CLR Table
write.table(d, file="test_clr.txt") # Output CLR Table

options(scipen=999) # to disable scientific notation
chr01_window <- sliding.window.transform(chr01, width = 10000, jump = 1000, whole.data = TRUE, type=2)
chr01_window <- sweeps.stats(chr01_window, freq.table=freq.table)
chr01_window@CLR
d <- cbind(chr01_window@region.names, chr01_window@CLR) # Make CLR Table
write.table(d, file="test_clr.txt") # Output CLR Table
```


### MINOR ALLELE FREQUENCIES ###
```{r}

## Get MAF for whole genome
onePop <- concatenate.regions(pv_data)
onePop <- detail.stats(onePop, site.spectrum = TRUE)
onePop@region.stats@minor.allele.freqs[[1]][1] # returns the MAF for the first SNP
hist(onePop@region.stats@minor.allele.freqs[[1]]) 
plot(density(onePop@region.stats@minor.allele.freqs[[1]]))








## Get MAF for genes
## Important for comparing Pv to Pf, since Pf only has good cov in coding regions
onePop_genes <- splitting.data(onePop, subsites = "gene")
onePop_genes <- detail.stats(onePop_genes, site.spectrum = TRUE)
onePop_genes@region.stats@minor.allele.freqs[1] # gives us the MAFs for the first gene
gene_mafs <- unlist(onePop_genes@region.stats@minor.allele.freqs) # gets rid of the list-of-lists structure
plot(density(gene_mafs))


num <- NULL

for (i in 1:length(onePop_genes@region.stats@minor.allele.freqs)) {

print i
  
}



## Define populations: BB, KP, and OM
indivs <- get.individuals(onePop)
threePops <- set.populations(onePop, list(
  indivs[[1]][1:69][grep("BB", indivs[[1]][1:length(indivs[[1]])])], 
  indivs[[1]][1:69][grep("KP", indivs[[1]][1:length(indivs[[1]])])], 
  indivs[[1]][1:69][grep("OM", indivs[[1]][1:length(indivs[[1]])])]))

threePops <- detail.stats(threePops, site.spectrum = TRUE)
threePops@region.stats@minor.allele.freqs[[1]][1]

## What a pain, it's intercalating the values from the three populations
```
